### PR TITLE
[match] Only encrypt and commit files if they changed

### LINKED ISSUE #8793: [match] Optimize match
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

Hello, we international company and having more > 2000 apps. 
When we release update match clone git and decrypt all profiles for one app! after check him and again encrypt! For this he spends a lot of time to update only 1 application. But apps > 2000....it is need amazing time. 
May be you can optimize match - that for 1 apps encrypted and decrypted only need profile and other don't touch. 

may be you can help me ?
may be you can add mode for match don't work with git? in realtime connect to apple find check download profile and sign app?)
may be add mode work only with filesystem without decrypt and without git?

### LINKED ISSUE #8917: Match repository getting huge!
Hello

We are using fastlane match for a lot of apps, and it seems to change every certificate every time match is run, so for example the last commit match did changed 384 files, because of this the repo gains about 5mb in size each time and we just hit 1gb.

Is there anything I can do to make it only push the ones it generates rather than the whole lot? Or is this to do with how the encryption works?

The only flag we are using is `shallow_clone`

### Environment

<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.12.4                                     |
| Ruby                        | 2.2.4                                       |
| Bundler?                    | false                                       |
| Git                         | git version 2.11.0 (Apple Git-81)           |
| Installation Source         | ~/.fastlane/bin/bundle/bin/fastlane         |
| Host                        | Mac OS X 10.12.4 (16E195)                   |
| Ruby Lib Dir                | ~/.fastlane/bin/bundle/lib                  |
| OpenSSL Version             | OpenSSL 1.0.2g  1 Mar 2016                  |
| Is contained                | false                                       |
| Is homebrew                 | true                                        |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 8.3                                         |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE | en_US.UTF-8 | ✅ |

### fastlane files:

**No Fastfile found**

<details><summary>`./fastlane/Appfile`</summary>

```ruby
app_identifier "net.xxxx.app01" # The bundle identifier of your app
apple_id "xxxxxx" # Your Apple email address
team_id "xxxx"  # Developer Portal Team ID
```
</details>

### fastlane gems

| Gem      | Version | Update-Status      |
| -------- | ------- | ------------------ |
| fastlane | 2.24.0  | 🚫 Update available |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| slack-notifier            | 1.5.1        |
| i18n                      | 0.7.0        |
| json                      | 1.8.1        |
| thread_safe               | 0.3.5        |
| tzinfo                    | 1.2.2        |
| minitest                  | 5.4.3        |
| activesupport             | 4.2.7.1      |
| CFPropertyList            | 2.3.5        |
| claide                    | 1.0.1        |
| colored                   | 1.2          |
| nanaimo                   | 0.2.3        |
| xcodeproj                 | 1.4.2        |
| rouge                     | 1.11.1       |
| xcpretty                  | 0.2.4        |
| terminal-notifier         | 1.7.1        |
| unicode-display_width     | 1.1.3        |
| terminal-table            | 1.7.3        |
| plist                     | 3.2.0        |
| public_suffix             | 2.0.5        |
| addressable               | 2.5.0        |
| multipart-post            | 2.0.0        |
| word_wrap                 | 1.0.0        |
| tty-screen                | 0.5.0        |
| babosa                    | 1.0.2        |
| highline                  | 1.7.8        |
| commander-fastlane        | 4.4.4        |
| excon                     | 0.54.0       |
| faraday                   | 0.11.0       |
| unf_ext                   | 0.0.7.2      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20161129 |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| fastimage                 | 2.0.1        |
| gh_inspector              | 1.0.3        |
| uber                      | 0.0.15       |
| representable             | 2.3.0        |
| retriable                 | 2.1.0        |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| hurley                    | 0.2          |
| little-plugger            | 1.1.4        |
| multi_json                | 1.12.1       |
| logging                   | 2.1.0        |
| jwt                       | 1.5.6        |
| memoist                   | 0.15.0       |
| os                        | 0.9.6        |
| signet                    | 0.7.3        |
| googleauth                | 0.5.1        |
| httpclient                | 2.8.3        |
| google-api-client         | 0.9.26       |
| mini_magick               | 4.5.1        |
| multi_xml                 | 0.6.0        |
| rubyzip                   | 1.2.0        |
| security                  | 0.1.3        |
| xcpretty-travis-formatter | 0.0.4        |
| dotenv                    | 2.2.0        |
| bundler                   | 1.14.3       |
| faraday_middleware        | 0.11.0.1     |
</details>

*generated on:* **2017-04-19**
</details>

### PR DESCRIPTION
**Pull request by KrauseFx**
This is work in progress, tests are failing still

**Before**: https://github.com/KrauseFx/certs/commit/d13691626cae6cb476c0aa41ef82cf9803d3feff
**Now**: https://github.com/KrauseFx/certs/commit/2edbbcd2ff9f55352898791bdd56bd2e4dfc2524

So much cleaner :) 

- Fixes https://github.com/fastlane/fastlane/issues/8917
- Fixes https://github.com/fastlane/fastlane/issues/8793
- Replaces, closes and fixes https://github.com/fastlane/fastlane/pull/8919

### CODE REVIEW

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: match/lib/match/git_helper.rb [hunk 1]
```diff
@@ -80,21 +81,45 @@ def self.generate_commit_message(params)
...
         Encrypt.new.encrypt_repo(path: path, git_url: git_url)
-        File.write("match_version.txt", Fastlane::VERSION) # unencrypted
-
         commands = []
-        commands << "git add -A"
+
+        if files_to_commmit.count > 0 # e.g. for nuke this is treated differently
+          if !File.exist?(MATCH_VERSION_FILE_NAME) || File.read(MATCH_VERSION_FILE_NAME) != Fastlane::VERSION.to_s
+            files_to_commmit << MATCH_VERSION_FILE_NAME
+            File.write(MATCH_VERSION_FILE_NAME, Fastlane::VERSION) # stored unencrypted
```
Comment (ohayon) [inline match/lib/match/git_helper.rb]: Is all of this code in this `if` block to just keep track of what version of _match_ was used to encrypt the stuff?

**Review by ohayon — APPROVED** [review]
Just one question, but looks 💯

File: match/lib/match/git_helper.rb (hunk 1 above)
Comment (KrauseFx) [reply match/lib/match/git_helper.rb]: Yep 👍

**Pull request merged**